FEA Implement an optimized sliding window matmul kernel

fcharras opened the pull request:
Necessary step before going forward with the KNN. This kernel will enable computing pairwise distances.

ogrisel commented:
Nice, quick benchmark vs `np.dot` (backend by OpenBLAS SGEMM) on random `float32` arrays, e.g.:

```python
import numpy as np
n, m, k = 2048, 2048, 2048
dtype = np.float32

A = np.random.randn(n, k).astype(dtype)
B = np.random.randn(k, m).astype(dtype)
C = np.empty(shape=(n, m), dtype=dtype)
%timeit np.dot(A, B, out=C)
```

fcharras commented:
Numpy
```
70 ms ± 2.2 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

sklearn_numba_dpex

```
158 ms ± 442 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

We're only twice slower :grin:

ogrisel commented:
sycl-blas has some Python wrapper here: https://github.com/codeplaysoftware/sycl-blas/blob/master/python_generator/py_gen_blas_gemm_launcher.py

ogrisel commented:
Actually this doesn't look like a wrapper. I am not sure what's the purpose of this code. Maybe a code generator.

jjerphan commented:
Thanks for the benchmarks, @fcharras. :)

I would be in favor of reviewing and merging this PR if there's no downside: this implementation works, is efficient, and can be made even more efficient then if we want it to.

What do you think?

fcharras commented:
Yes IMO it can be reviewed

fcharras commented:
`dpnp.matmul` on GPU provides 10* speedup on numpy. And 20* speedup on the kernel in this PR. Will probably just use `dpnp.matmul` instead.

jjerphan commented on sklearn_numba_dpex/common/matmul.py:
```diff
@@ -0,0 +1,277 @@
+
+    It is expected to take one scalar argument and return one scalar value.
+
+    Likewise, if `multiply_fn` is not None it will be used in place of the scalar
+    multiply operator . It is expected to take two scalar argument and return one scalar
+    value.
```
```suggestion
    Likewise, if `multiply_fn` is not None it will be used in place of the scalar
    multiply operator. It is expected to take two scalar arguments and return one
    scalar value.
```

jjerphan commented on sklearn_numba_dpex/common/matmul.py:
```diff
@@ -0,0 +1,277 @@
+        X_sliding_window = dpex.local.array(shape=window_shape, dtype=dtype)
+        Y_t_sliding_window = dpex.local.array(shape=window_shape, dtype=dtype)
+
+        # Index of the column of the sliding window that the current work item will be
+        # responsible for loading. The "sliding" is materialize by incrementing this
+        # index by `square_window_side` at each iteration of the main loop.
```
```suggestion
        # Index of the column of the sliding window that the current work item will be
        # responsible for loading. The "sliding" is materialized by incrementing this
        # index by `square_window_side` at each iteration of the main loop.
```

jjerphan commented on sklearn_numba_dpex/common/matmul.py:
```diff
@@ -0,0 +1,277 @@
+        # optimize RW memory access patterns, it's best to ensure that contiguous work
+        # items read contiguous memory slots in both shared and global memory. This is
+        # straightforward to ensure for windows on `X`, but requires the following
+        # remapped index for `Y_T`:
```
```suggestion
        # remapped index for `Y_t`:
```

jjerphan commented on sklearn_numba_dpex/common/matmul.py:
```diff
@@ -0,0 +1,277 @@
+
+            dpex.barrier(dpex.LOCAL_MEM_FENCE)
+
+        # Write result
```
```suggestion
        # Save the value of the scalar product of the result_row_idx-th row of X
        # and of the result_col_idx-th column of Y.
```

jjerphan reviewed (commented):
A few comments. Do we need to introduce this kernel since `dnpn.matmul` exist?

fcharras commented:
> A few comments. Do we need to introduce this kernel since `dnpn.matmul` exist?

I wonder. I don't understand how the performance look so weak here, while our kmeans seems to work well, at least it compares as one could expect with cpu implementation.

I suppose the performances will not be as bad with typical input dimension for KNN (with `n_cols` = `n_features` is much lesser than the other two dimensions), and then it might actually be better to use this kernel because we can fuse `multiply_fn` and `elementwise_fn` to compute pairwise distances in one kernel call.

fcharras commented:
I'm currently working on an update on this. I think it's actually possible to achieve near dpnp performance with numba_dpex.

fcharras commented:
Also I've opened a discussion at `numba_dpex` see https://github.com/IntelPython/numba-dpex/issues/984

ogrisel commented on sklearn_numba_dpex/common/matmul.py:
```diff
@@ -0,0 +1,286 @@
+zero_idx = np.int64(0)
+
+
+@lru_cache
+def make_matmul_2d_kernel(
```
I am not sure we need the full GEMM API in most cases and specializing it for simple MM might be more efficient for the most common cases (while keeping the code simpler).

I think a full GEMM kernel (or family of kernels to be specialized for subcases) can be done upstream in numba_dpex if needed.

ogrisel commented on sklearn_numba_dpex/common/matmul.py:
```diff
@@ -0,0 +1,814 @@
+        # of results covered by this work group
+        group_row_idx = group_idx // global_grid_n_cols
+        group_col_idx = group_idx % global_grid_n_cols
+        # TODO: improve cache locality ?
```
nit

```suggestion
        # TODO: improve cache locality?
```

ogrisel commented on sklearn_numba_dpex/common/matmul.py:
```diff
@@ -0,0 +1,814 @@
+    # Setting it too high might pressure local or private memory too much for it to be
+    # worth.
+    arithmetic_intensity_multiplier_X = 4
+    arithmetic_intensity_multiplier_Y = 2
```
Let's make those kwargs of `make_matmul_2d_kernel` with the same default values so that it will be possible in the future to do run some kind of auto-tuning grid search.

ogrisel reviewed (approved):
I won't review the core of the kernel itself but +1 for merge with stronger non-regression tests as explained below (at least a few more test cases with larger shapes and maybe non-integer, positive and negative input values).

fcharras commented on sklearn_numba_dpex/common/tests/test_matmul.py:
```diff
@@ -0,0 +1,109 @@
+    X_shape, Y_shape = test_input_shapes
+
+    X = _arange_reshaped(X_shape, dtype)
+    Y = _arange_reshaped(Y_shape, dtype)
```
It doesn't cost much more to have both the `_arange_reshaped` and `random` tests so I can add both. (I think `_arange_reshaped` is enough, but above that it is handier for debug than random values so I'd like to keep it along)

I think the boundary effects are already all covered by this set of test cases (the inputs are smalls but the tests also use small values for `work_group_size` / `sub_group_size`)

fcharras commented:
Latest push should have addressed comments, this version is mergeable IMO.

- [x] improved tests
- [x] removed WIP work on cache locality (and pushed it in the matmul_kernel_wip on this repo)
- [x] exposed all performance parameters
